### PR TITLE
Nuke GuardDuty (#175)

### PR DESCRIPTION
**Pull request by jdpleiness**
* Add support for removing GuardDuty as requested in #175
* Update README with new functionality

### CODE REVIEW

File: aws/gd.go [hunk 1]
```diff
@@ -0,0 +1,74 @@
...
+	ctx := context.Background()
+	var detectors []*string
+
+	err := svc.ListDetectorsPagesWithContext(ctx, &guardduty.ListDetectorsInput{},
+		func(d *guardduty.ListDetectorsOutput, lastPage bool) bool {
+			for _, detector := range d.DetectorIds {
+				detectors = append(detectors, detector)
+			}
+			return true
+		})
```
Comment (brikis98) [inline aws/gd.go:26]: Don't you need to use the `lastPage` variable here?

File: aws/gd.go [hunk 2]
```diff
@@ -0,0 +1,74 @@
...
+)
+
+// getAllGuardDutyDetectors returns active GuardDuty detectors in a given region.
+func getAllGuardDutyDetectors(session *session.Session, excludeAfter time.Time) ([]*string, error) {
+	svc := guardduty.New(session)
+
+	ctx := context.Background()
+	var detectors []*string
+
+	err := svc.ListDetectorsPagesWithContext(ctx, &guardduty.ListDetectorsInput{},
```
Comment (brikis98) [inline aws/gd.go:20]: Is GuardDuty enabled in every AWS region? If not, we'd need to handle the error when it's not enabled, and skip that region.

File: aws/gd_test.go [hunk 3]
```diff
@@ -0,0 +1,64 @@
...
+	return *detector.DetectorId
+}
+
+func TestNukeAllGuardDutyDetectors(t *testing.T) {
+	t.Parallel()
+
+	region, err := getRandomRegion()
+	if err != nil {
+		assert.Fail(t, errors.WithStackTrace(err).Error())
+	}
```
Comment (brikis98) [inline aws/gd_test.go:34]:
```suggestion
	require.NoError(t, err)
```

**Review by brikis98 — COMMENTED** [review]
Thanks for the PR!

File: aws/gd.go (hunk 1 above)
Comment (jdpleiness) [reply aws/gd.go:26]: `lastPage` is part of the method signature of `ListDetectorsPagesWithContext` and is used in that function as part of the paging process. [see here](https://docs.aws.amazon.com/sdk-for-go/api/service/guardduty/#GuardDuty.ListDetectorsPagesWithContext)

File: aws/gd.go [hunk 4]
```diff
@@ -0,0 +1,74 @@
...
+	"github.com/gruntwork-io/cloud-nuke/logging"
+	"github.com/gruntwork-io/go-commons/errors"
+)
+
+// getAllGuardDutyDetectors returns active GuardDuty detectors in a given region.
+func getAllGuardDutyDetectors(session *session.Session, excludeAfter time.Time) ([]*string, error) {
+	svc := guardduty.New(session)
+
+	ctx := context.Background()
+	var detectors []*string
```
[the remark this answers is not in the document]
Comment (jdpleiness) [reply aws/gd.go:18]: `detectorIds` is a better name, I will go ahead and change that. I also really didn't like using the slice of pointers either, but did so to try and fit with project code convention. Examples [here](https://github.com/gruntwork-io/cloud-nuke/blob/922a488ac17f4e8c845467e7dd29e91a192209f9/aws/ami.go#L26) as well as [here](https://github.com/gruntwork-io/cloud-nuke/blob/922a488ac17f4e8c845467e7dd29e91a192209f9/aws/lambda.go#L22). I'd be happy to change this as well though.

File: aws/gd.go (hunk 2 above)
Comment (jdpleiness) [reply aws/gd.go:20]: I might be misunderstanding what you mean by enabled, but the method does not return an error if there are no GuardDuty detectors in a region, only for service API and SDK errors.

File: aws/gd_test.go (hunk 3 above)
Comment (jdpleiness) [reply aws/gd_test.go:34]: This is great, I haven't had much experience with the testify package so I was unaware of this! I will go through and change these.

File: aws/gd_test.go [hunk 5]
```diff
@@ -0,0 +1,64 @@
...
+
+	assert.NotContains(t, awsgo.StringValueSlice(detectors), detectorID)
+
+	detectors, err = getAllGuardDutyDetectors(sess, time.Now().Add(1*time.Hour))
+	if err != nil {
+		assert.Fail(t, "Unable to list GuardDuty detector in region")
+	}
+
+	assert.Contains(t, awsgo.StringValueSlice(detectors), detectorID)
```
[the remark this answers is not in the document]
Comment (jdpleiness) [reply aws/gd_test.go:63]: The detectors are nuked via the preceding `defer` statement but a final check to make sure they are gone is a great addition.

**Comment by jdpleiness** [comment]
Updated the PR based on your feedback. Let me know if I missed anything.

**Comment by ina-stoyanova** [comment]
hey @jdpleiness thanks for raising this PR & addressing the feedback. Do you mind getting it up to speed with the latest from master, and we can hopefully help moving it forward faster?

**Comment by ekristen** [comment]
@ina-stoyanova I happened to implement my own before I realized there was an open PR and since this one is stale by a couple months, I went ahead and opened my PR https://github.com/gruntwork-io/cloud-nuke/pull/239

**Comment by jdpleiness** [comment]
@ekristen @ina-stoyanova I got sidetracked and forgot about this for a while. It should be back up to speed with master now if you would like to merge it. If you would like to use @ekristen's version of the feature instead, feel free to close close the PR. Either is fine with me.

**Comment by ina-stoyanova** [comment]
Hey @jdpleiness & @ekristen thank you both so much for the contributions! I'll take a look as soon as possible, and we can see how we go about merging both! :)

**Comment by ina-stoyanova** [comment]
>Hey @jdpleiness & @ekristen thank you both so much for the contributions! I'll take a look as soon as possible, and we can see how we go about merging both! :)

just a heads up - that's looking like next week :)

**Comment by ina-stoyanova** [comment]
@jdpleiness thanks for raising this PR! I already commented on @ekristen's, but to be transparent - we're currently ramping up someone to help us maintain this repo and hopefully, we'll take a look at merging your PRs and adding the support for Guardduty as soon as possible. This is more likely to be pushed back by another week while the ramping up is still going on.

**Comment by jdpleiness** [comment]
@ina-stoyanova Are there any updates on this? I know these things take time, but it's been 7 months and other PRs have been merged as well as a new release. It would be great to have support for GuardDuty. I don't see any updates on @ekristen's PR either at this point.

**Pull request closed without merging**